ports-of-call/portable_arrays.hpp

```diff
@@ -24,438 +24,275 @@
 //  [N4xN3xN2xN1] are accessed as:  A(n,k,j,i) = A[i + N1*(j + N2*(k + N3*n))]
 //  NOTE THE TRAILING INDEX INSIDE THE PARENTHESES IS INDEXED FASTEST
 
+#include "array.hpp"
 #include "portability.hpp"
+#include "span.hpp"
+#include "utility/array_algo.hpp"
+#include "utility/index_algo.hpp"
 #include <algorithm>
 #include <assert.h>
 #include <cstddef> // size_t
 #include <cstring> // memset()
+#include <functional>
+#include <type_traits>
 #include <utility> // swap()
 
-template <typename T>
+namespace PortsOfCall {
+// maximum number of dimensions
+constexpr std::size_t MAXDIM = 6;
+
+namespace detail {
+// convert `index_sequence` values to constant
+template <std::size_t I, std::size_t V>
+constexpr std::size_t to_const = V;
+
+} // namespace detail
+
+// if C++20, we can use std::array without worry on device
+// else use a hand-rolled array
+#if __cplusplus == 202002L
+template <class T, auto N>
+using Array = std::array<T, N>;
+#else
+template <class T, auto N>
+using Array = PortsOfCall::array<T, N>;
+#endif // __cplusplus
+
+template <auto N>
+using IArray = Array<std::size_t, N>;
+
+template <typename T, std::size_t D = MAXDIM>
 class PortableMDArray {
  public:
-  static constexpr int MAXDIM = 6;
+  using this_type = PortableMDArray<T, D>;
```

Comment:
We should have some introspection on the maximum dimension of `PortableMDArray`. Can we add a member field that reports `D`?